File: community/cypher/src/main/scala/org/neo4j/cypher/internal/parser/v1_6/Predicates.scala
/**
 * Copyright (c) 2002-2012 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.parser.v1_6

import org.neo4j.graphdb.Direction
import org.neo4j.cypher.SyntaxException
import org.neo4j.cypher.internal.commands._

trait Predicates extends Base with Expressions with ReturnItems {
  def predicate: Parser[Predicate] = (
    isNull 
      | isNotNull 
      | orderedComparison 
      | not 
      | notEquals 
      | equals 
      | regexp 
      | hasProperty 
      | parens(predicate) 
      | sequencePredicate 
      | hasRelationshipTo 
      | hasRelationship
      | aggregateFunctionNames ~> parens(expression) ~> failure("aggregate functions can not be used in the WHERE clause")
    
    ) * (
    ignoreCase("and") ^^^ {
      (a: Predicate, b: Predicate) => And(a, b)
    } |
      ignoreCase("or") ^^^ {
        (a: Predicate, b: Predicate) => Or(a, b)
      }
    )

  def regexp: Parser[Predicate] = expression ~ "=~" ~ (regularLiteral | expression) ^^ {
    case a ~ "=~" ~ b => RegularExpression(a, b)
  }

  def hasProperty: Parser[Predicate] = property ^^ ( prop => Has(prop.asInstanceOf[Property]))

  def sequencePredicate: Parser[Predicate] = allInSeq | anyInSeq | noneInSeq | singleInSeq

  def symbolIterablePredicate: Parser[(Expression, String, Predicate)] =
    (identity ~ ignoreCase("in") ~ expression ~ ignoreCase("where")  ~ predicate ^^ {    case symbol ~ in ~ iterable ~ where ~ klas => (iterable, symbol, klas)  }
      |identity ~> ignoreCase("in") ~ expression ~> failure("expected where"))


  def allInSeq: Parser[Predicate] = ignoreCase("all") ~> parens(symbolIterablePredicate) ^^ (x => AllInIterable(x._1, x._2, x._3))

  def anyInSeq: Parser[Predicate] = ignoreCase("any") ~> parens(symbolIterablePredicate) ^^ (x => AnyInIterable(x._1, x._2, x._3))

  def noneInSeq: Parser[Predicate] = ignoreCase("none") ~> parens(symbolIterablePredicate) ^^ (x => NoneInIterable(x._1, x._2, x._3))

  def singleInSeq: Parser[Predicate] = ignoreCase("single") ~> parens(symbolIterablePredicate) ^^ (x => SingleInIterable(x._1, x._2, x._3))

  def equals: Parser[Predicate] = expression ~ "=" ~ expression ^^ {
    case l ~ "=" ~ r => (l, r) match {
      case (a:Nullable, b:Nullable) => NullablePredicate(Equals(a,b), Seq((a,true), (b,true)))
      case (a:Nullable, b) =>  NullablePredicate(Equals(a,b), Seq((a,true)))
      case (a,b:Nullable)=>  NullablePredicate(Equals(a,b), Seq((b,true)))
      case (a,b)=> Equals(l, r)
    }
  }

  def notEquals: Parser[Predicate] = expression ~ ("!=" | "<>") ~ expression ^^ { case l ~ wut ~ r => Not(Equals(l, r)) }

  def orderedComparison: Parser[Predicate] = (lessThanOrEqual | greaterThanOrEqual | lessThan | greaterThan)

  def lessThan: Parser[Predicate] = expression ~ "<" ~ expression ^^ { case l ~ "<" ~ r => LessThan(l, r) }

  def greaterThan: Parser[Predicate] = expression ~ ">" ~ expression ^^ { case l ~ ">" ~ r => GreaterThan(l, r) }

  def lessThanOrEqual: Parser[Predicate] = expression ~ "<=" ~ expression ^^ { case l ~ "<=" ~ r => LessThanOrEqual(l, r) }

  def greaterThanOrEqual: Parser[Predicate] = expression ~ ">=" ~ expression ^^ { case l ~ ">=" ~ r => GreaterThanOrEqual(l, r) }

  def not: Parser[Predicate] = ignoreCase("not") ~ "(" ~ predicate ~ ")" ^^ { case not ~ "(" ~ inner ~ ")" => Not(inner) }

  def expressionOrEntity = expression | entity

  def isNull: Parser[Predicate] = expressionOrEntity <~ ignoreCase("is null") ^^ (x => IsNull(x))

  def isNotNull: Parser[Predicate] = expressionOrEntity <~ ignoreCase("is not null") ^^ (x => Not(IsNull(x)))

  def hasRelationshipTo: Parser[Predicate] = expressionOrEntity ~ relInfo ~ expressionOrEntity ^^ { case a ~ rel ~ b => HasRelationshipTo(a, b, rel._1, rel._2) }

  def hasRelationship: Parser[Predicate] = expressionOrEntity ~ relInfo <~ "()" ^^ { case a ~ rel  => HasRelationship(a, rel._1, rel._2) }

  def relInfo: Parser[(Direction, Seq[String])] = opt("<") ~ "-" ~ opt("[:" ~> identity <~ "]") ~ "-" ~ opt(">") ^^ {
    case Some("<") ~ "-" ~ relType ~ "-" ~ Some(">") => throw new SyntaxException("Can't be connected both ways.", "query", 666)
    case Some("<") ~ "-" ~ relType ~ "-" ~ None => (Direction.INCOMING, relType.toSeq)
    case None ~ "-" ~ relType ~ "-" ~ Some(">") => (Direction.OUTGOING, relType.toSeq)
    case None ~ "-" ~ relType ~ "-" ~ None => (Direction.BOTH, relType.toSeq)
  }
}

File: community/common/src/main/java/org/neo4j/function/Predicates.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.function;

import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.neo4j.function.ThrowingPredicate.throwingPredicate;
import static org.neo4j.function.ThrowingSupplier.throwingSupplier;

import java.time.Clock;
import java.util.Arrays;
import java.util.Objects;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.locks.LockSupport;
import java.util.function.BooleanSupplier;
import java.util.function.IntPredicate;
import java.util.function.LongPredicate;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * Constructors for basic {@link Predicate} types
 */
public final class Predicates {
    public static final IntPredicate ALWAYS_TRUE_INT = v -> true;
    public static final IntPredicate ALWAYS_FALSE_INT = v -> false;
    public static final LongPredicate ALWAYS_TRUE_LONG = v -> true;
    public static final LongPredicate ALWAYS_FALSE_LONG = v -> false;

    private static final int DEFAULT_POLL_INTERVAL = 20;
    private static final int DEFAULT_TIMEOUT_MS = 20_000;

    private Predicates() {}

    public static <T> Predicate<T> alwaysTrue() {
        return x -> true;
    }

    public static <T> Predicate<T> alwaysFalse() {
        return x -> false;
    }

    public static <T> Predicate<T> notNull() {
        return Objects::nonNull;
    }

    @SafeVarargs
    public static <T> Predicate<T> all(final Predicate<T>... predicates) {
        return all(Arrays.asList(predicates));
    }

    public static <T> Predicate<T> all(final Iterable<Predicate<T>> predicates) {
        return item -> {
            for (Predicate<T> predicate : predicates) {
                if (!predicate.test(item)) {
                    return false;
                }
            }
            return true;
        };
    }

    @SafeVarargs
    public static <T> Predicate<T> any(final Predicate<T>... predicates) {
        return any(Arrays.asList(predicates));
    }

    public static <T> Predicate<T> any(final Iterable<Predicate<T>> predicates) {
        return item -> {
            for (Predicate<T> predicate : predicates) {
                if (predicate.test(item)) {
                    return true;
                }
            }
            return false;
        };
    }

    public static <T> Predicate<T> instanceOf(final Class<?> clazz) {
        return clazz::isInstance;
    }

    public static <T> Predicate<T> instanceOfAny(final Class<?>... classes) {
        return item -> {
            if (item != null) {
                for (Class<?> clazz : classes) {
                    if (clazz.isInstance(item)) {
                        return true;
                    }
                }
            }
            return false;
        };
    }

    public static <TYPE> TYPE await(
            Supplier<TYPE> supplier,
            Predicate<TYPE> predicate,
            long timeout,
            TimeUnit timeoutUnit,
            long pollInterval,
            TimeUnit pollUnit)
            throws TimeoutException {
        return awaitEx(supplier::get, predicate::test, timeout, timeoutUnit, pollInterval, pollUnit);
    }

    public static <TYPE> TYPE await(
            Supplier<TYPE> supplier, Predicate<TYPE> predicate, long timeout, TimeUnit timeoutUnit)
            throws TimeoutException {
        return awaitEx(throwingSupplier(supplier), throwingPredicate(predicate), timeout, timeoutUnit);
    }

    public static <TYPE, EXCEPTION extends Exception> TYPE awaitEx(
            ThrowingSupplier<TYPE, EXCEPTION> supplier,
            ThrowingPredicate<TYPE, EXCEPTION> predicate,
            long timeout,
            TimeUnit timeoutUnit,
            long pollInterval,
            TimeUnit pollUnit)
            throws TimeoutException, EXCEPTION {
        Suppliers.ThrowingCapturingSupplier<TYPE, EXCEPTION> composed = Suppliers.compose(supplier, predicate);
        awaitEx(composed, timeout, timeoutUnit, pollInterval, pollUnit);
        return composed.lastInput();
    }

    public static <TYPE, EXCEPTION extends Exception> TYPE awaitEx(
            ThrowingSupplier<TYPE, ? extends EXCEPTION> supplier,
            ThrowingPredicate<TYPE, ? extends EXCEPTION> predicate,
            long timeout,
            TimeUnit timeoutUnit)
            throws TimeoutException, EXCEPTION {
        Suppliers.ThrowingCapturingSupplier<TYPE, EXCEPTION> composed = Suppliers.compose(supplier, predicate);
        awaitEx(composed, timeout, timeoutUnit);
        return composed.lastInput();
    }

    public static void await(BooleanSupplier condition) throws TimeoutException {
        awaitEx(condition::getAsBoolean, DEFAULT_TIMEOUT_MS, MILLISECONDS);
    }

    public static void await(BooleanSupplier condition, long timeout, TimeUnit unit) throws TimeoutException {
        awaitEx(condition::getAsBoolean, timeout, unit);
    }

    public static <EXCEPTION extends Exception> void awaitEx(
            ThrowingSupplier<Boolean, EXCEPTION> condition, long timeout, TimeUnit unit)
            throws TimeoutException, EXCEPTION {
        awaitEx(condition, timeout, unit, DEFAULT_POLL_INTERVAL, MILLISECONDS);
    }

    public static void await(
            BooleanSupplier condition, long timeout, TimeUnit timeoutUnit, long pollInterval, TimeUnit pollUnit)
            throws TimeoutException {
        awaitEx(condition::getAsBoolean, timeout, timeoutUnit, pollInterval, pollUnit);
    }

    public static <EXCEPTION extends Exception> void awaitEx(
            ThrowingSupplier<Boolean, EXCEPTION> condition,
            long timeout,
            TimeUnit unit,
            long pollInterval,
            TimeUnit pollUnit)
            throws TimeoutException, EXCEPTION {
        if (!tryAwaitEx(condition, timeout, unit, pollInterval, pollUnit)) {
            throw new TimeoutException(
                    "Waited for " + timeout + " " + unit + ", but " + condition + " was not accepted.");
        }
    }

    public static <EXCEPTION extends Exception> boolean tryAwaitEx(
            ThrowingSupplier<Boolean, EXCEPTION> condition,
            long timeout,
            TimeUnit timeoutUnit,
            long pollInterval,
            TimeUnit pollUnit)
            throws EXCEPTION {
        return tryAwaitEx(condition, timeout, timeoutUnit, pollInterval, pollUnit, Clock.systemUTC());
    }

    public static <EXCEPTION extends Exception> boolean tryAwaitEx(
            ThrowingSupplier<Boolean, EXCEPTION> condition,
            long timeout,
            TimeUnit timeoutUnit,
            long pollInterval,
            TimeUnit pollUnit,
            Clock clock)
            throws EXCEPTION {
        long deadlineMillis = clock.millis() + timeoutUnit.toMillis(timeout);
        long pollIntervalNanos = pollUnit.toNanos(pollInterval);

        do {
            if (condition.get()) {
                return true;
            }
            LockSupport.parkNanos(pollIntervalNanos);
        } while (clock.millis() < deadlineMillis);
        return false;
    }

    public static void awaitForever(BooleanSupplier condition, long checkInterval, TimeUnit unit) {
        long sleep = unit.toNanos(checkInterval);
        do {
            if (condition.getAsBoolean()) {
                return;
            }
            LockSupport.parkNanos(sleep);
        } while (true);
    }

    @SafeVarargs
    public static <T> Predicate<T> in(final T... allowed) {
        return in(Arrays.asList(allowed));
    }

    public static <T> Predicate<T> in(final Iterable<T> allowed) {
        return item -> {
            for (T allow : allowed) {
                if (allow.equals(item)) {
                    return true;
                }
            }
            return false;
        };
    }

    public static IntPredicate any(int[] values) {
        return v -> {
            for (int value : values) {
                if (v == value) {
                    return true;
                }
            }
            return false;
        };
    }
}
